lib/spack/llnl/util/tty/log.py

```diff
@@ -88,43 +129,122 @@ def __init__(self, stream):
+
+        """
+        with self.defer_signals():
+            # query terminal flags and fg/bg status
+            flags = self._get_canon_echo_flags()
+            bg = self._is_background()
+
+            # restore sanity if any flags are amiss
+            if not bg and any(flags):    # fg, but input not enabled
+                self._enable_keyboard_input()
+            elif bg and not all(flags):  # bg, but input enabled
+                self._restore_input()
+
+            # reinstall TSTP handler on CONT (see below where TSTP removes it)
+            if signum == signal.SIGCONT:
+                signal.signal(signal.SIGTSTP, self._bg_fg_handler)
+
+            # python can't block signals, so we reinstall the default TSTP
+            # handler here, which actually stops the process on kill
+            if signum == signal.SIGTSTP:
+                signal.signal(
+                    signal.SIGTSTP, self.old_handlers[signal.SIGTSTP])
+                os.kill(os.getpid(), signal.SIGTSTP)
 
     def __enter__(self):
-        """Enable immediate keypress input on stream.
+        """Enable immediate keypress input, while this process is foreground.
 
         If the stream is not a TTY or the system doesn't support termios,
         do nothing.
         """
         self.old_cfg = None
+        self.old_handlers = {}
 
         # Ignore all this if the input stream is not a tty.
         if not self.stream or not self.stream.isatty():
-            return
+            return self
 
-        # If this fails, self.old_cfg will remain None
-        if termios and not _is_background_tty():
-            # save old termios settings
-            old_cfg = termios.tcgetattr(self.stream)
+        if termios:
+            # save old termios settings to restore later
+            self.old_cfg = termios.tcgetattr(self.stream)
 
-            try:
-                # create new settings with canonical input and echo
-                # disabled, so keypresses are immediate & don't echo.
-                self.new_cfg = termios.tcgetattr(self.stream)
-                self.new_cfg[3] &= ~termios.ICANON
-                self.new_cfg[3] &= ~termios.ECHO
+            # Install a signal handler to disable/enable keyboard input
+            # when the process moves between foreground and background.
+            for signum in self.signals:
+                self.old_handlers[signum] = signal.signal(
+                    signum, self._bg_fg_handler)
+
+            # add an atexit handler to ensure the terminal is restored
+            atexit.register(self._restore_input)
 
-                # Apply new settings for terminal
-                termios.tcsetattr(self.stream, termios.TCSADRAIN, self.new_cfg)
-                self.old_cfg = old_cfg
+            # this itimer handles cases where the OS doesn't notify us of
+            # a background or foreground change.
+            signal.setitimer(signal.ITIMER_REAL, 1e-3, 1e-3)
```

Comment:
I think this can be `1e-1`: one tenth of a second is sufficiently responsive for humans.

Reply:
I could do 1e-2, but I think 1e-1 is too coarse -- humans can hit keys faster than that.

Reply:
ok looked at this in more details -- 1e-1 is probably a good idea.  1e-3 makes the daemon use ~16% CPU ; 1e-2 is 2-4% and 1e-1 is ~0.5% CPU.  And I can't tell a difference with `v`.

Reply:
updated to 1e-1